packages/flutter_tools/lib/src/android/native_assets.dart

```diff
@@ -0,0 +1,204 @@
+Future<Uri?> dryRunNativeAssetsAndroid({
+  required NativeAssetsBuildRunner buildRunner,
+  required Uri projectUri,
+  bool flutterTester = false,
+  required FileSystem fileSystem,
+}) async {
+  if (!await nativeBuildRequired(buildRunner)) {
+    return null;
+  }
+
+  final Uri buildUri_ = nativeAssetsBuildUri(projectUri, OS.android);
+  final Iterable<Asset> nativeAssetPaths =
+      await dryRunNativeAssetsAndroidInternal(
+          fileSystem, projectUri, buildRunner);
+  final Uri nativeAssetsUri =
+      await writeNativeAssetsYaml(nativeAssetPaths, buildUri_, fileSystem);
+  return nativeAssetsUri;
+}
+
+Future<Iterable<Asset>> dryRunNativeAssetsAndroidInternal(
+  FileSystem fileSystem,
+  Uri projectUri,
+  NativeAssetsBuildRunner buildRunner,
+) async {
+  const OS targetOS = OS.android;
+
+  globals.logger.printTrace('Dry running native assets for $targetOS.');
+  final List<Asset> nativeAssets = (await buildRunner.dryRun(
+    linkModePreference: LinkModePreference.dynamic,
+    targetOS: targetOS,
+    workingDirectory: projectUri,
+    includeParentEnvironment: true,
+  ))
+      .assets;
+  ensureNoLinkModeStatic(nativeAssets);
+  globals.logger.printTrace('Dry running native assets for $targetOS done.');
+  final Map<Asset, Asset> assetTargetLocations =
+      _assetTargetLocations(nativeAssets);
+  final Iterable<Asset> nativeAssetPaths = assetTargetLocations.values;
+  return nativeAssetPaths;
+}
+
+/// Builds native assets.
+Future<(Uri? nativeAssetsYaml, List<Uri> dependencies)>
+    buildNativeAssetsAndroid({
+  required NativeAssetsBuildRunner buildRunner,
+  required Iterable<AndroidArch> androidArchs,
+  required Uri projectUri,
+  required BuildMode buildMode,
+  String? codesignIdentity,
+  Uri? yamlParentDirectory,
+  required FileSystem fileSystem,
+}) async {
+  const OS targetOS = OS.android;
+  final Uri buildUri_ = nativeAssetsBuildUri(projectUri, targetOS);
+  if (!await nativeBuildRequired(buildRunner)) {
+    final Uri nativeAssetsYaml = await writeNativeAssetsYaml(
+      <Asset>[],
+      yamlParentDirectory ?? buildUri_,
+      fileSystem,
+    );
+    return (nativeAssetsYaml, <Uri>[]);
+  }
+
+  final List<Target> targets = androidArchs.map(_getNativeTarget).toList();
+  final native_assets_cli.BuildMode buildModeCli =
+      nativeAssetsBuildMode(buildMode);
+
+  globals.logger
+      .printTrace('Building native assets for $targets $buildModeCli.');
+  final List<Asset> nativeAssets = <Asset>[];
+  final Set<Uri> dependencies = <Uri>{};
+  for (final Target target in targets) {
+    final BuildResult result = await buildRunner.build(
+      linkModePreference: LinkModePreference.dynamic,
+      target: target,
+      buildMode: buildModeCli,
+      workingDirectory: projectUri,
+      includeParentEnvironment: true,
+      cCompilerConfig: await buildRunner.ndkCCompilerConfig,
+      // TODO(dacoharkes): Get this value from somewhere.
+      targetAndroidNdkApi: 30,
```

Comment:
Should this be using https://github.com/flutter/flutter/blob/8860dd21d8e91deba17ca3349e57e572e1567b75/packages/flutter_tools/gradle/src/main/groovy/flutter.groovy#L58-L62 ?

Reply:
@stuartmorgan The NDK version and API level are not the same.

* API level is tied to Android version: https://developer.android.com/tools/releases/platforms
* NDK version can target certain Android versions.

We currently pin the NDK to 23.1.xxx (Used to be 21.1.xxx [flutter.dev/go/android-ndk-version](http://flutter.dev/go/android-ndk-version)).
The minimum Android version Flutter currently supports is Android 21. https://docs.flutter.dev/reference/supported-platforms

@reidbaker We discussed adding a minimum Android API level to the `pubspec.yaml`.

Thinking about that a bit more, should we only add that to the root package, and then pass the same API version to all `build.dart` invocations? And then the `build.dart` scripts can check for the Android API version and fail compilation if it's too low.

The alternative would require us to parse the `pubspec.yaml` of every package before invoking `build.dart`. Moreover, this would not require the root-package (the actual flutter app) to specify a version, which would be weird. It's the developer making an app that decides what the lowest Android API version is to target.

Also, we should probably be making the android NDK version optional in the pubspec. We should probably default to the minimum supported API version: 21. (Or we could even not provide a config option at all for now, and always pass 21.)

Reply:
Apps specify the minimum SDK level already: https://docs.flutter.dev/deployment/android#under-the-defaultconfig-block

The code defaults to best effort min SDK version, not lowest supported min SDK version:

https://github.com/flutter/flutter/blob/69c2ab4fb046a9bee1ec020d468807f358eddb4b/packages/flutter_tools/gradle/src/main/groovy/flutter.groovy#L46-L47

https://github.com/flutter/flutter/blob/69c2ab4fb046a9bee1ec020d468807f358eddb4b/packages/flutter_tools/lib/src/android/gradle_utils.dart#L42

We encourage users to set the value in the `build.gradle` of their app. Setting it in the pubspec for an app then becomes inconsequential (especially if not running via `flutter run`), unless build.gradle starts reading from the pubspec.

Alternatively, flutter_tools could try to read the value from `build.gradle`, but this seems fragile.

If we do specify an minSdkVersion in the pubspec of packages with build.dart, then we'd need to need to check in flutter.groovy that none of these versions is newer than the version in the `build.gradle` of the app. However, I believe this to be wrong approach. This would not enable to take advantage of optimizations enabled in newer Android versions.

(Side note: It seems that we can ignore target SDK version. Min SDK is the lowest Android an app will run on, but targetSdk is what it is designed to run on. The target SDK version seems to be inconsequential unless devs call the runtime function to get the current SDK version. https://developer.android.com/ndk/guides/sdk-versions#targetsdkversion The NDK documentation doesn't give any info about using the target version when doing command line compilations, only the min SDK version. https://developer.android.com/ndk/guides/other_build_systems)